test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndTransactionOrderTest.java

```diff
@@ -0,0 +1,245 @@
+        CompletableFuture<Void> writer3 = startWriter("3", clientFactory, done);
+        CompletableFuture<Void> writer4 = startWriter("4", clientFactory, done);
+
+        // perform multiple scale stream operations so that rolling transactions may happen
+        Stream s = new StreamImpl("test", "test");
+        Map<Double, Double> map = new HashMap<>();
+        map.put(0.0, 1.0);
+        ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+
+        controller.scaleStream(s, Collections.singletonList(0L), map, executor).getFuture().get();
+
+        controller.scaleStream(s, Collections.singletonList(StreamSegmentNameUtils.computeSegmentId(1, 1)), map, executor).getFuture().get();
+
+        controller.scaleStream(s, Collections.singletonList(StreamSegmentNameUtils.computeSegmentId(2, 2)), map, executor).getFuture().get();
+
+        // stop writers
+        done.set(true);
+
+        CompletableFuture.allOf(writer1, writer2, writer3, writer4).join();
+
+        // wait for all transactions to commit
+        Futures.allOf(eventToTxnMap.entrySet().stream()
+                                   .map(x -> waitTillCommitted(controller, s, x.getValue(), uncommitted)).collect(Collectors.toList())).join();
+
+        assertTrue(uncommitted.isEmpty());
+        // read all events using a single reader and verify the order
+
+        List<Triple<Integer, UUID, String>> eventOrder = new LinkedList<>();
+        // create a reader
+        while (!eventToTxnMap.isEmpty()) {
+            EventRead<Integer> integerEventRead = reader.readNextEvent(SECONDS.toMillis(60));
+            if (integerEventRead.getEvent() != null) {
+                int event1 = integerEventRead.getEvent();
+                UUID txnId = eventToTxnMap.remove(event1);
+                String writerId = txnToWriter.get(txnId);
+                UUID first = writersList.get(writerId).remove(0);
+                eventOrder.add(new ImmutableTriple<>(event1, txnId, writerId));
+                assertEquals(first, txnId);
+            }
+        }
+    }
+
+    private CompletableFuture<Void> waitTillCommitted(Controller controller, Stream s, UUID key, ConcurrentSet<UUID> uncommitted) {
+        AtomicBoolean committed = new AtomicBoolean(false);
+        AtomicInteger counter = new AtomicInteger(0);
+        // check 6 times with 5 second gap until transaction is committed. if it is not committed, declare it uncommitted
+        return Futures.loop(() -> !committed.get() && counter.getAndIncrement() < 5,
+                () -> Futures.delayedFuture(() -> controller.checkTransactionStatus(s, key)
+                                                            .thenAccept(status -> {
+                                                                committed.set(status.equals(Transaction.Status.COMMITTED));
+                                                            }), 5000, executor), executor)
```

Comment:
the problem with that is since we are creating very large number of transcations in this test (around 1000s), we would hammer our threadpool with too many requests if we do it more aggressively.